File: src/main/java/kirjanpito/ui/DocumentUIBuilder.java
package kirjanpito.ui;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.GridLayout;
import java.awt.Insets;
import java.awt.event.ActionListener;
import java.awt.event.KeyAdapter;
import java.awt.event.KeyEvent;

import javax.swing.AbstractAction;
import javax.swing.Action;
import javax.swing.BorderFactory;
import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JSeparator;
import javax.swing.JTextField;
import javax.swing.KeyStroke;
import javax.swing.SwingConstants;

import kirjanpito.models.TextFieldWithLockIcon;
import kirjanpito.ui.resources.Resources;

/**
 * Hallinnoi DocumentFrame:n UI-komponenttien luomista.
 * 
 * Tämä luokka eriytetään DocumentFrame:stä Phase 7 -refaktoroinnin osana.
 * Se vastaa kaikkien UI-komponenttien luomisesta ja konfiguroinnista.
 * 
 * @since 2.2.4
 */
public class DocumentUIBuilder {
    
    private final UICallbacks callbacks;
    
    /**
     * Callback-rajapinta DocumentFrame:lle.
     */
    public interface UICallbacks {
        /** Kutsutaan kun tositenumero-kenttä muuttuu */
        void onNumberFieldChanged();
        
        /** Kutsutaan kun päivämäärä-kenttä muuttuu */
        void onDateFieldChanged();
        
        /** Kutsutaan kun haku-painiketta klikataan */
        void onSearchButtonClicked();
        
        /** Kutsutaan kun haku-paneeli suljetaan */
        void onSearchPanelClosed();
        
        /** Palauttaa addEntry-actionin */
        Action getAddEntryAction();
        
        /** Palauttaa searchListener-actionin */
        ActionListener getSearchListener();
    }
    
    /**
     * Komponenttien viitteet.
     */
    public static class UIComponents {
        public TextFieldWithLockIcon numberTextField;
        public DateTextField dateTextField;
        public JLabel debitTotalLabel;
        public JLabel creditTotalLabel;
        public JLabel differenceLabel;
        public JPanel searchPanel;
        public JTextField searchPhraseTextField;
        public JLabel documentLabel;
        public JLabel periodLabel;
        public JLabel documentTypeLabel;
        public JLabel backupStatusLabel;
    }
    
    private final UIComponents components;
    
    /**
     * Konstruktori.
     * 
     * @param callbacks Callback-rajapinta DocumentFrame:lle
     */
    public DocumentUIBuilder(UICallbacks callbacks) {
        this.callbacks = callbacks;
        this.components = new UIComponents();
    }
    
    /**
     * Palauttaa luodut komponentit.
     */
    public UIComponents getComponents() {
        return components;
    }
    
    /**
     * Luo tekstikenttäpaneelin (tositenumero ja päivämäärä).
     * 
     * @param container Paneeli, johon komponentit lisätään
     */
    public void createTextFieldPanel(JPanel container) {
        GridBagConstraints c;
        JPanel panel = new JPanel();
        container.add(panel, BorderLayout.PAGE_START);
        panel.setLayout(new GridLayout(0, 2));

        JPanel left = new JPanel();
        left.setLayout(new GridBagLayout());
        panel.add(left);

        /* Lisätään paneeliin tositenumeronimiö ja -tekstikenttä. */
        JLabel numberLabel = new JLabel("Tositenumero");
        c = new GridBagConstraints();
        c.anchor = GridBagConstraints.WEST;
        c.insets = new Insets(8, 8, 8, 4);
        left.add(numberLabel, c);

        components.numberTextField = new TextFieldWithLockIcon();
        components.numberTextField.setCaret(new javax.swing.text.DefaultCaret());
        components.numberTextField.addKeyListener(new KeyAdapter() {
            public void keyReleased(KeyEvent e) {
                if (components.numberTextField.isEditable()) {
                    callbacks.onNumberFieldChanged();
                }
            }
        });

        components.numberTextField.getInputMap().put(
                KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), "transferFocus");

        components.numberTextField.getActionMap().put("transferFocus", new AbstractAction() {
            private static final long serialVersionUID = 1L;
            @Override
            public void actionPerformed(java.awt.event.ActionEvent e) {
                components.numberTextField.transferFocus();
            }
        });

        c = new GridBagConstraints();
        c.insets = new Insets(8, 4, 8, 4);
        c.fill = GridBagConstraints.HORIZONTAL;
        c.weightx = 1.0;
        left.add(components.numberTextField, c);

        c.fill = GridBagConstraints.VERTICAL;
        c.weightx = 0.0;
        left.add(new JSeparator(SwingConstants.VERTICAL), c);

        JPanel right = new JPanel();
        right.setLayout(new GridBagLayout());
        panel.add(right);

        /* Lisätään paneeliin päivämääränimiö ja -tekstikenttä. */
        JLabel dateLabel = new JLabel("Päivämäärä");
        c = new GridBagConstraints();
        c.anchor = GridBagConstraints.WEST;
        c.insets = new Insets(8, 8, 8, 4);
        right.add(dateLabel, c);

        components.dateTextField = new DateTextField();
        components.dateTextField.setCaret(new javax.swing.text.DefaultCaret());
        components.dateTextField.addKeyListener(new KeyAdapter() {
            public void keyReleased(KeyEvent e) {
                if (components.dateTextField.isEditable() && Character.isDigit(e.getKeyChar())) {
                    callbacks.onDateFieldChanged();
                }
            }
        });

        components.dateTextField.getInputMap().put(
                KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), "addEntry");

        components.dateTextField.getActionMap().put("addEntry", callbacks.getAddEntryAction());

        components.dateTextField.getInputMap().put(
                KeyStroke.getKeyStroke(KeyEvent.VK_DOWN, 0), "firstEntry");

        components.dateTextField.getActionMap().put("firstEntry", new AbstractAction() {
            private static final long serialVersionUID = 1L;

            @Override
            public void actionPerformed(java.awt.event.ActionEvent e) {
                components.dateTextField.transferFocus();
                // Note: First entry selection logic will be handled by DocumentFrame
            }
        });

        c = new GridBagConstraints();
        c.insets = new Insets(8, 4, 8, 8);
        c.fill = GridBagConstraints.HORIZONTAL;
        c.weightx = 1.0;
        right.add(components.dateTextField, c);
    }
    
    /**
     * Luo summarivin (debet/kredit yhteenvedot).
     * 
     * @param container Paneeli, johon rivi lisätään
     */
    public void createTotalRow(JPanel container) {
        GridBagLayout layout = new GridBagLayout();
        GridBagConstraints c = new GridBagConstraints();
        c.insets = new Insets(2, 2, 2, 6);
        c.anchor = GridBagConstraints.WEST;

        JPanel panel = new JPanel(layout);
        panel.setBorder(BorderFactory.createEmptyBorder(2, 2, 5, 2));
        container.add(panel);

        components.debitTotalLabel = new JLabel("0,00");
        Dimension minSize = components.debitTotalLabel.getMinimumSize();
        components.debitTotalLabel.setPreferredSize(new Dimension(80, minSize.height));
        components.creditTotalLabel = new JLabel("0,00");
        components.creditTotalLabel.setPreferredSize(new Dimension(80, minSize.height));
        components.differenceLabel = new JLabel("0,00");
        components.differenceLabel.setPreferredSize(new Dimension(80, minSize.height));

        panel.add(new JLabel("Debet yht."), c);
        panel.add(components.debitTotalLabel, c);
        panel.add(new JLabel("Kredit yht."), c);
        panel.add(components.creditTotalLabel, c);
        panel.add(new JLabel("Erotus"), c);
        c.weightx = 1.0;
        panel.add(components.differenceLabel, c);
    }
    
    /**
     * Luo hakupalkin.
     * 
     * @param container Paneeli, johon hakupalkki lisätään
     */
    public void createSearchBar(JPanel container) {
        GridBagLayout layout = new GridBagLayout();
        GridBagConstraints c = new GridBagConstraints();
        c.insets = new Insets(8, 5, 8, 5);
        c.anchor = GridBagConstraints.WEST;

        JPanel panel = components.searchPanel = new JPanel(layout);
        panel.setBorder(BorderFactory.createBevelBorder(javax.swing.border.BevelBorder.LOWERED));
        panel.setVisible(false);
        container.add(panel);

        JTextField textField = components.searchPhraseTextField = new JTextField();
        textField.setCaret(new javax.swing.text.DefaultCaret());
        textField.addKeyListener(new KeyAdapter() {
            public void keyPressed(KeyEvent e) {
                if (e.getKeyCode() == KeyEvent.VK_ENTER) {
                    callbacks.onSearchButtonClicked();
                    e.consume();
                }
            }
        });
        c.weightx = 1.0;
        c.fill = GridBagConstraints.HORIZONTAL;
        panel.add(textField, c);

        JButton button = new JButton("Etsi", new ImageIcon(
                Resources.loadAsImage("find-16x16.png")));

        button.addActionListener(e -> callbacks.onSearchButtonClicked());
        button.setMnemonic('H');
        c.weightx = 0.0;
        c.fill = GridBagConstraints.BOTH;
        panel.add(button, c);

        button = new JButton(new ImageIcon(Resources.loadAsImage("close-16x16.png")));
        button.addActionListener(callbacks.getSearchListener());
        panel.add(button, c);
    }
    
    /**
     * Luo tilarivin (status bar).
     * 
     * @param frame JFrame, johon status bar lisätään
     */
    public void createStatusBar(javax.swing.JFrame frame) {
        JPanel statusBarPanel = new JPanel(new BorderLayout());
        components.documentLabel = new JLabel();
        components.documentLabel.setBorder(new javax.swing.border.EtchedBorder());
        components.documentLabel.setPreferredSize(new Dimension(150, 0));
        components.periodLabel = new JLabel(" ");
        components.periodLabel.setBorder(new javax.swing.border.EtchedBorder());
        components.documentTypeLabel = new JLabel();
        components.documentTypeLabel.setBorder(new javax.swing.border.EtchedBorder());
        components.documentTypeLabel.setPreferredSize(new Dimension(200, 0));
        
        // Backup-indikaattori (Word AutoSave -tyylinen)
        components.backupStatusLabel = new JLabel();
        components.backupStatusLabel.setBorder(new javax.swing.border.EtchedBorder());
        components.backupStatusLabel.setPreferredSize(new Dimension(120, 0));
        components.backupStatusLabel.setHorizontalAlignment(SwingConstants.CENTER);
        
        // Backup-indikaattoria klikkaamalla avautuu asetukset
        components.backupStatusLabel.addMouseListener(new java.awt.event.MouseAdapter() {
            @Override
            public void mouseClicked(java.awt.event.MouseEvent e) {
                // Callback will be set by DocumentFrame
            }
        });
        components.backupStatusLabel.setCursor(new java.awt.Cursor(java.awt.Cursor.HAND_CURSOR));
        components.backupStatusLabel.setToolTipText("Klikkaa muuttaaksesi varmuuskopiointiasetuksia");
        
        JPanel rightPanel = new JPanel(new BorderLayout());
        rightPanel.add(components.backupStatusLabel, BorderLayout.WEST);
        rightPanel.add(components.documentTypeLabel, BorderLayout.CENTER);
        
        statusBarPanel.add(components.documentLabel, BorderLayout.WEST);
        statusBarPanel.add(components.periodLabel, BorderLayout.CENTER);
        statusBarPanel.add(rightPanel, BorderLayout.EAST);
        frame.add(statusBarPanel, BorderLayout.PAGE_END);
    }
}

File: src/main/kotlin/kirjanpito/ui/DocumentUIBuilder.kt
package kirjanpito.ui

import kirjanpito.models.TextFieldWithLockIcon
import kirjanpito.ui.resources.Resources
import java.awt.BorderLayout
import java.awt.Cursor
import java.awt.Dimension
import java.awt.GridBagConstraints
import java.awt.GridBagLayout
import java.awt.GridLayout
import java.awt.Insets
import java.awt.event.ActionListener
import java.awt.event.KeyAdapter
import java.awt.event.KeyEvent
import javax.swing.*
import javax.swing.border.BevelBorder
import javax.swing.border.EtchedBorder
import javax.swing.text.DefaultCaret

/**
 * Hallinnoi DocumentFrame:n UI-komponenttien luomista.
 *
 * Tämä luokka eriytetään DocumentFrame:stä Phase 7 -refaktoroinnin osana.
 * Se vastaa kaikkien UI-komponenttien luomisesta ja konfiguroinnista.
 * Migrated from Java to Kotlin for better code conciseness and null-safety.
 *
 * @since 2.2.4
 */
class DocumentUIBuilder(private val callbacks: UICallbacks) {

    /**
     * Callback-rajapinta DocumentFrame:lle.
     */
    interface UICallbacks {
        /** Kutsutaan kun tositenumero-kenttä muuttuu */
        fun onNumberFieldChanged()

        /** Kutsutaan kun päivämäärä-kenttä muuttuu */
        fun onDateFieldChanged()

        /** Kutsutaan kun haku-painiketta klikataan */
        fun onSearchButtonClicked()

        /** Kutsutaan kun haku-paneeli suljetaan */
        fun onSearchPanelClosed()

        /** Palauttaa addEntry-actionin */
        fun getAddEntryAction(): Action

        /** Palauttaa searchListener-actionin */
        fun getSearchListener(): ActionListener
    }

    /**
     * Komponenttien viitteet.
     */
    class UIComponents {
        @JvmField var numberTextField: TextFieldWithLockIcon? = null
        @JvmField var dateTextField: DateTextField? = null
        @JvmField var debitTotalLabel: JLabel? = null
        @JvmField var creditTotalLabel: JLabel? = null
        @JvmField var differenceLabel: JLabel? = null
        @JvmField var searchPanel: JPanel? = null
        @JvmField var searchPhraseTextField: JTextField? = null
        @JvmField var documentLabel: JLabel? = null
        @JvmField var periodLabel: JLabel? = null
        @JvmField var documentTypeLabel: JLabel? = null
        @JvmField var backupStatusLabel: JLabel? = null
    }

    val components = UIComponents()

    /**
     * Luo tekstikenttäpaneelin (tositenumero ja päivämäärä).
     *
     * @param container Paneeli, johon komponentit lisätään
     */
    fun createTextFieldPanel(container: JPanel) {
        val panel = JPanel().apply {
            layout = GridLayout(0, 2)
        }
        container.add(panel, BorderLayout.PAGE_START)

        // Left panel - Tositenumero
        val left = JPanel(GridBagLayout()).apply {
            // Tositenumero label
            add(JLabel("Tositenumero"), GridBagConstraints().apply {
                anchor = GridBagConstraints.WEST
                insets = Insets(8, 8, 8, 4)
            })

            // Tositenumero field
            this@DocumentUIBuilder.components.numberTextField = TextFieldWithLockIcon().apply {
                caret = DefaultCaret()
                addKeyListener(object : KeyAdapter() {
                    override fun keyReleased(e: KeyEvent) {
                        if (isEditable) {
                            callbacks.onNumberFieldChanged()
                        }
                    }
                })

                inputMap.put(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), "transferFocus")
                actionMap.put("transferFocus", object : AbstractAction() {
                    override fun actionPerformed(e: java.awt.event.ActionEvent) {
                        transferFocus()
                    }
                })
            }

            add(this@DocumentUIBuilder.components.numberTextField!!, GridBagConstraints().apply {
                insets = Insets(8, 4, 8, 4)
                fill = GridBagConstraints.HORIZONTAL
                weightx = 1.0
            })

            // Separator
            add(JSeparator(SwingConstants.VERTICAL), GridBagConstraints().apply {
                fill = GridBagConstraints.VERTICAL
                weightx = 0.0
            })
        }
        panel.add(left)

        // Right panel - Päivämäärä
        val right = JPanel(GridBagLayout()).apply {
            // Päivämäärä label
            add(JLabel("Päivämäärä"), GridBagConstraints().apply {
                anchor = GridBagConstraints.WEST
                insets = Insets(8, 8, 8, 4)
            })

            // Päivämäärä field
            this@DocumentUIBuilder.components.dateTextField = DateTextField().apply {
                caret = DefaultCaret()
                addKeyListener(object : KeyAdapter() {
                    override fun keyReleased(e: KeyEvent) {
                        if (isEditable && e.keyChar.isDigit()) {
                            callbacks.onDateFieldChanged()
                        }
                    }
                })

                inputMap.put(KeyStroke.getKeyStroke(KeyEvent.VK_ENTER, 0), "addEntry")
                actionMap.put("addEntry", callbacks.getAddEntryAction())

                inputMap.put(KeyStroke.getKeyStroke(KeyEvent.VK_DOWN, 0), "firstEntry")
                actionMap.put("firstEntry", object : AbstractAction() {
                    override fun actionPerformed(e: java.awt.event.ActionEvent) {
                        transferFocus()
                        // Note: First entry selection logic will be handled by DocumentFrame
                    }
                })
            }

            add(this@DocumentUIBuilder.components.dateTextField!!, GridBagConstraints().apply {
                insets = Insets(8, 4, 8, 8)
                fill = GridBagConstraints.HORIZONTAL
                weightx = 1.0
            })
        }
        panel.add(right)
    }

    /**
     * Luo summarivin (debet/kredit yhteenvedot).
     *
     * @param container Paneeli, johon rivi lisätään
     */
    fun createTotalRow(container: JPanel) {
        val panel = JPanel(GridBagLayout()).apply {
            border = BorderFactory.createEmptyBorder(2, 2, 5, 2)
        }
        container.add(panel)

        val constraints = GridBagConstraints().apply {
            insets = Insets(2, 2, 2, 6)
            anchor = GridBagConstraints.WEST
        }

        components.debitTotalLabel = JLabel("0,00").apply {
            preferredSize = Dimension(80, minimumSize.height)
        }
        components.creditTotalLabel = JLabel("0,00").apply {
            preferredSize = Dimension(80, minimumSize.height)
        }
        components.differenceLabel = JLabel("0,00").apply {
            preferredSize = Dimension(80, minimumSize.height)
        }

        panel.add(JLabel("Debet yht."), constraints)
        panel.add(components.debitTotalLabel, constraints)
        panel.add(JLabel("Kredit yht."), constraints)
        panel.add(components.creditTotalLabel, constraints)
        panel.add(JLabel("Erotus"), constraints)
        panel.add(components.differenceLabel, constraints.apply { weightx = 1.0 })
    }

    /**
     * Luo hakupalkin.
     *
     * @param container Paneeli, johon hakupalkki lisätään
     */
    fun createSearchBar(container: JPanel) {
        val panel = JPanel(GridBagLayout()).apply {
            border = BorderFactory.createBevelBorder(BevelBorder.LOWERED)
            isVisible = false
        }
        components.searchPanel = panel
        container.add(panel)

        val constraints = GridBagConstraints().apply {
            insets = Insets(8, 5, 8, 5)
            anchor = GridBagConstraints.WEST
        }

        // Search text field
        this@DocumentUIBuilder.components.searchPhraseTextField = JTextField().apply {
            caret = DefaultCaret()
            addKeyListener(object : KeyAdapter() {
                override fun keyPressed(e: KeyEvent) {
                    if (e.keyCode == KeyEvent.VK_ENTER) {
                        callbacks.onSearchButtonClicked()
                        e.consume()
                    }
                }
            })
        }

        panel.add(components.searchPhraseTextField, constraints.apply {
            weightx = 1.0
            fill = GridBagConstraints.HORIZONTAL
        })

        // Search button
        val searchButton = JButton("Etsi", ImageIcon(Resources.loadAsImage("find-16x16.png"))).apply {
            addActionListener { callbacks.onSearchButtonClicked() }
            mnemonic = 'H'.code
        }

        panel.add(searchButton, constraints.apply {
            weightx = 0.0
            fill = GridBagConstraints.BOTH
        })

        // Close button
        val closeButton = JButton(ImageIcon(Resources.loadAsImage("close-16x16.png"))).apply {
            addActionListener(callbacks.getSearchListener())
        }

        panel.add(closeButton, constraints)
    }

    /**
     * Luo tilarivin (status bar).
     *
     * @param frame JFrame, johon status bar lisätään
     */
    fun createStatusBar(frame: JFrame) {
        val statusBarPanel = JPanel(BorderLayout())

        this@DocumentUIBuilder.components.documentLabel = JLabel().apply {
            border = EtchedBorder()
            preferredSize = Dimension(150, 0)
        }

        components.periodLabel = JLabel(" ").apply {
            border = EtchedBorder()
        }

        this@DocumentUIBuilder.components.documentTypeLabel = JLabel().apply {
            border = EtchedBorder()
            preferredSize = Dimension(200, 0)
        }

        // Backup-indikaattori (Word AutoSave -tyylinen)
        this@DocumentUIBuilder.components.backupStatusLabel = JLabel().apply {
            border = EtchedBorder()
            preferredSize = Dimension(120, 0)
            horizontalAlignment = SwingConstants.CENTER
            cursor = Cursor(Cursor.HAND_CURSOR)
            toolTipText = "Klikkaa muuttaaksesi varmuuskopiointiasetuksia"

            addMouseListener(object : java.awt.event.MouseAdapter() {
                override fun mouseClicked(e: java.awt.event.MouseEvent) {
                    // Callback will be set by DocumentFrame
                }
            })
        }

        val rightPanel = JPanel(BorderLayout()).apply {
            add(this@DocumentUIBuilder.components.backupStatusLabel!!, BorderLayout.WEST)
            add(this@DocumentUIBuilder.components.documentTypeLabel!!, BorderLayout.CENTER)
        }

        statusBarPanel.apply {
            add(this@DocumentUIBuilder.components.documentLabel!!, BorderLayout.WEST)
            add(this@DocumentUIBuilder.components.periodLabel!!, BorderLayout.CENTER)
            add(rightPanel, BorderLayout.EAST)
        }

        frame.add(statusBarPanel, BorderLayout.PAGE_END)
    }
}
